File: src/Python/gauss_seidel.py
import numpy as np
from tridiag import estTridiagonale


def gauss_seidel(A, b, x0, eps, itmax):
    return gauss_seidel_tridiag(A, b, x0, eps, itmax) if estTridiagonale(A) else gauss_seidel_(A, b, x0, eps, itmax)


def gauss_seidel_tridiag(A, b, x0, eps, itmax):
    it = 0
    err = np.linalg.norm(b - A.dot(x0))
    x_old = x0
    n = len(A)
    x = np.zeros(n)
    while err > eps and it < itmax:
        S = A[0][1] * x_old[1]
        x[0] = (b[0] - S) / A[0][0]
        for i in range(1, n-1):
            S = A[i][i-1] * x[i-1] + A[i][i+1] * x_old[i+1]
            x[i] = (b[i] - S) / A[i][i]
        i += 1
        S = A[i][i-1] * x[i-1]
        x[i] = (b[i] - S) / A[i][i]
        x_old = x
        err = np.linalg.norm(b - A.dot(x))
        it += 1
    return x, it


def gauss_seidel_(A, b, x0, eps, itmax):
    it = 0
    err = np.linalg.norm(b - A.dot(x0))
    x_old = x0
    n = len(A)
    x = np.zeros(n)
    while err > eps and it < itmax:
        for i in range(n):
            S = 0
            for j in range(i):
                S += A[i][j] * x[j]
            for j in range(i + 1, n):
                S += A[i][j] * x_old[j]
            x[i] = (b[i] - S) / A[i][i]
        x_old = x
        err = np.linalg.norm(b - A.dot(x))
        it += 1
    return x, it

File: src/C/gauss_seidel.c
#include <stdlib.h>
#include "matrice.h"
#include "jgs.h"

double *gauss_seidel_tridiag(double *A, double *b, double *x0, double eps, int itmax, int n)
{
    double *x_old, *x, err, S;
    int it, i;
    it = 0;
    err = norm_2(soustraction(b, produit(A, x0, n), n), n);
    x_old = x0;
    x = (double *)calloc(n, sizeof(double));
    while (err > eps && it < itmax)
    {
        S = *(A + 1) * *(x_old + 1);
        *x = (*b - S) / *A;
        for (i = 0; i < n - 1; i++)
        {
            S = *(A + i * n + i - 1) * *(x + i - 1) + *(A + i * n + i + 1) * *(x_old + i + 1);
            *(x + i) = (*(b + i) - S) / *(A + i * n + i);
        }
        S = *(A + i * n + i - 1) * *(x + i - 1);
        *(x + i) = (*(b + i) - S) / *(A + i * n + i);
        x_old = x;
        err = norm_2(soustraction(b, produit(A, x, n), n), n);
        it++;
    }
    return x;
}

double *gauss_seidel_(double *A, double *b, double *x0, double eps, int itmax, int n)
{
    double *x_old, *x, err, S;
    int it, i, j;
    it = 0;
    err = norm_2(soustraction(b, produit(A, x0, n), n), n);
    x_old = x0;
    x = (double *)calloc(n, sizeof(double));
    while (err > eps && it < itmax)
    {
        for (i = 0; i < n; i++)
        {
            S = 0;
            for (j = 0; j < i; j++)
                S += *(A + i * n + j) * *(x_old + j);
            for (j = i + 1; j < n; j++)
                S += *(A + i * n + j) * *(x_old + j);
            *(x + i) = (*(b + i) - S) / *(A + i * n + i);
        }
        x_old = x;
        err = norm_2(soustraction(b, produit(A, x, n), n), n);
        it++;
    }
    return x;
}

double *gauss_seidel(double *A, double *b, double *x0, double eps, int itmax, int n)
{
    return estTridiagonale(A, n) ? gauss_seidel_tridiag(A, b, x0, eps, itmax, n) : gauss_seidel_(A, b, x0, eps, itmax, n);
}
